File: reference/lib/src/reference_counter.dart
import 'package:flutter/foundation.dart';

import '../reference.dart';

/// Contains lifecycle callbacks for [OwnerCounter].
class OwnerCounterLifecycleListener {
  const OwnerCounterLifecycleListener({
    @required this.onCreate,
    @required this.onDispose,
  })  : assert(onCreate != null),
        assert(onDispose != null);

  /// Called when owner count reaches 1 and
  final void Function(LocalReference localReference) onCreate;
  final void Function(
    LocalReference localReference,
    RemoteReference remoteReference,
  ) onDispose;
}

class OwnerCounter {
  OwnerCounter(
    this.lifecycleListener, [
    int initialOwnerCount = 0,
  ])  : assert(initialOwnerCount == null || initialOwnerCount >= 0),
        _ownerCount = initialOwnerCount ?? 0;

  final OwnerCounterLifecycleListener lifecycleListener;

  int _ownerCount;
  int get ownerCount => _ownerCount;

  void increment(LocalReference localReference) {
    _ownerCount++;
    if (ownerCount == 1) {
      lifecycleListener?.onCreate(localReference);
    }
  }

  void decrement(
    LocalReference localReference,
    RemoteReference remoteReference,
  ) {
    assert(ownerCount > 0,
        '`release()` was called without calling `retain()` first. In other words, `release()` was called while `referenceCount == 0`. Reference count = $_ownerCount.');

    _ownerCount--;
    if (ownerCount == 0) {
      lifecycleListener?.onDispose(localReference, remoteReference);
    }
  }
}

File: reference/android/src/main/java/github/penguin/reference/reference/ReferenceCounter.java
package github.penguin.reference.reference;

public class ReferenceCounter {
  public interface LifecycleListener {
    void onCreate(final String referenceId, final ReferenceManager.ReferenceHolder holder);

    void onDispose(final String referenceId, final ReferenceManager.ReferenceHolder holder);
  }

  private int referenceCount = 0;
  public final LifecycleListener lifecycleListener;

  public ReferenceCounter(final LifecycleListener lifecycleListener) {
    if (lifecycleListener == null) throw new IllegalArgumentException();
    this.lifecycleListener = lifecycleListener;
  }

  public ReferenceCounter(final LifecycleListener lifecycleListener, int initialReferenceCount) {
    this(lifecycleListener);
    if (initialReferenceCount < 0) throw new IllegalArgumentException();
    referenceCount = initialReferenceCount;
  }

  public void retain(final String referenceId, final ReferenceManager.ReferenceHolder holder) {
    referenceCount++;
    if (referenceCount == 1) lifecycleListener.onCreate(referenceId, holder);
  }

  public void release(final String referenceId, final ReferenceManager.ReferenceHolder holder) {
    if (referenceCount < 1) {
      final String message =
          "`release()` was called without calling `retain()` first. In other words, `release()` was called while `referenceCount == 0`. Reference count ="
              + referenceCount;
      throw new IllegalStateException("");
    }

    referenceCount--;
    if (referenceCount == 0) lifecycleListener.onDispose(referenceId, holder);
  }

  public int getReferenceCount() {
    return referenceCount;
  }
}
